dataset/shape.cpp

```diff
@@ -141,4 +141,271 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
+  const auto front_flat = reshape(front, {{to_dim, front.dims().volume()}});
+  const auto back_flat = reshape(back, {{to_dim, back.dims().volume()}});
+  // Check that bin edges can be joined together
+  if (front_flat.slice({to_dim, 1, front.dims().volume()}) !=
+      back_flat.slice({to_dim, 0, back.dims().volume() - 1}))
+    throw except::BinEdgeError(
+        "Flatten: the bin edges cannot be joined together.");
+
+  const auto base = var.slice({bin_edge_dim, 0, data_shape});
+  return concatenate(
+      reshape(base, flatten_dims(base.dims(), from_dims, to_dim)),
+      back_flat.slice({to_dim, back.dims().volume() - 1}), to_dim);
+}
+
+/// Check if one of the from_dims is a bin edge
+Dim bin_edge_in_from_dims(const VariableConstView &var,
+                          const Dimensions &array_dims,
+                          const Dimensions &from_dims) {
+  for (const auto dim : from_dims.labels())
+    if (is_bin_edges(var, array_dims, dim))
+      return dim;
+  return Dim::Invalid;
+}
+
+} // end anonymous namespace
+
+/// Split a single dimension into multiple dimensions:
+/// ['x': 6] -> ['y': 2, 'z': 3]
+DataArray split(const DataArrayConstView &a, const Dim from_dim,
+                const Dimensions &to_dims) {
+  // Make sure that new dims do not already exist in data dimensions,
+  // apart apart from the old dim (i.e. old dim can be re-used)
+  const auto &old_dims = a.dims();
+  for (const auto dim : to_dims.labels())
+    if (old_dims.contains(dim) && dim != from_dim)
+      throw except::DimensionError(
+          "Split: new dimensions cannot contain labels that "
+          "already exist in the DataArray.");
+
+  auto reshaped =
+      DataArray(reshape(a.data(), split_dims(old_dims, from_dim, to_dims)));
+
+  for (auto &&[name, coord] : a.coords()) {
+    if (is_bin_edges(coord, old_dims, from_dim))
+      reshaped.coords().set(name, split_bin_edge(coord, from_dim, to_dims));
+    else
+      reshaped.coords().set(
+          name, reshape(coord, split_dims(coord.dims(), from_dim, to_dims)));
+  }
+
+  for (auto &&[name, attr] : a.attrs())
+    if (is_bin_edges(attr, old_dims, from_dim))
+      reshaped.attrs().set(name, split_bin_edge(attr, from_dim, to_dims));
+    else
+      reshaped.attrs().set(
+          name, reshape(attr, split_dims(attr.dims(), from_dim, to_dims)));
+
+  // Note that we assume bin-edge masks do not exist
+  for (auto &&[name, mask] : a.masks())
+    reshaped.masks().set(
+        name, reshape(mask, split_dims(mask.dims(), from_dim, to_dims)));
+
+  return reshaped;
+}
+
+/// Flatten multiple dimensions into a single dimension:
+/// ['y', 'z'] -> ['x']
+DataArray flatten(const DataArrayConstView &a,
+                  const std::vector<Dim> &from_labels, const Dim to_dim) {
+  const auto &old_dims = a.dims();
+  for (const auto dim : from_labels)
+    if (!old_dims.contains(dim))
+      throw except::DimensionError(
+          "Flatten: dimension to be flattened not found in DataArray.");
```

Comment:
Please check that the dims validation functions are now doing what you would like

Reply:
Looks good.
It would be nice to have the dim labels in the error messages though. Maybe we can leave that for later, we have a lot of error messages to improve and should maybe use libfmt to make this easier.

Reply:
Good suggestion. I can add this while I am at it. It is often the kind of thing i point out when i review and then i find myself not doing in my own PRs.